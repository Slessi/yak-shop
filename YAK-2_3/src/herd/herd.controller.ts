import { Controller, Get, Param, ParseIntPipe } from '@nestjs/common';
import { HerdService } from './herd.service';

@Controller('herd')
export class HerdController {
  constructor(private herdService: HerdService) {}

  @Get(':T')
  getHerd(@Param('T', ParseIntPipe) T: number) {
    const data = this.herdService.getHerd(T);

    return {
      herd: data.map((d) => ({
        name: d.name,
        age: d.age,
        'age-last-shaved': d.ageLastShaved,
      })),
    };
  }
}
