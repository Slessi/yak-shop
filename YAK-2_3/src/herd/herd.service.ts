import { Injectable } from '@nestjs/common';
import { round } from 'src/round';

export interface LabYak {
  name: string;
  age: number;
  sex: string;
}

export interface LabYakResult extends LabYak {
  milk: number;
  skins: number;
  ageLastShaved: number;
}

@Injectable()
export class HerdService {
  public getHerd(totalDays: number) {
    const results = this.getData().map<LabYakResult>((l) => ({
      ...l,
      milk: 0,
      skins: 0,
      ageLastShaved: 0,
    }));

    let currentDay = 0;

    while (currentDay < totalDays) {
      results.forEach((result) => {
        // D = age in days
        const D = result.age * 100;

        // A LabYak dies the day he turns 10.
        if (result.age === 10) {
          return;
        }

        // Each day a LabYak produces 50-D*0.03 liters of milk
        result.milk += 50 - D * 0.03;

        if (this.canShaveLabYak(result, currentDay)) {
          result.skins += 1;
          result.ageLastShaved = result.age;
        }

        result.age = round(result.age + 0.01);
      });

      currentDay++;
    }

    return results.map((r) => ({
      ...r,
      milk: round(r.milk),
    }));
  }

  private canShaveLabYak(
    { age, ageLastShaved }: LabYakResult,
    currentDay: number,
  ) {
    // A yak can be first shaven when he is 1 year.
    if (age < 1) {
      return false;
    }

    // The moment you open up the YakShop webshop will be day 0, and all yaks are eligible to be shaven
    if (currentDay === 0) {
      return true;
    }

    // At most every 8+age (D*0.01) days you can again shave a LabYak
    return (age - ageLastShaved) * 100 > 8 + age;
  }

  // TODO: Get this from where ...?
  private getData(): LabYak[] {
    return [
      { name: 'Betty-1', age: 4, sex: 'f' },
      { name: 'Betty-2', age: 8, sex: 'f' },
      { name: 'Betty-3', age: 9.5, sex: 'f' },
    ];
  }
}
