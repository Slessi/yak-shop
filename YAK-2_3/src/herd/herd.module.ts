import { Module } from '@nestjs/common';
import { HerdController } from './herd.controller';
import { HerdService } from './herd.service';

@Module({
  controllers: [HerdController],
  providers: [HerdService],
  exports: [HerdService],
})
export class HerdModule {}
