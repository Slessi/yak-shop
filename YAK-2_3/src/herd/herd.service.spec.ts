import { HerdService } from './herd.service';

describe('HerdService', () => {
  it.each<number>([13, 14, 200])(
    '%# - calculates expected values for %i days',
    (T) => {
      expect(new HerdService().getHerd(T)).toMatchSnapshot();
    },
  );
});
