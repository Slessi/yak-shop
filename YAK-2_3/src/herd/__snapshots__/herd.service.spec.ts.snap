// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HerdService 0 - calculates expected values for 13 days 1`] = `
Array [
  Object {
    "age": 4.13,
    "ageLastShaved": 4,
    "milk": 491.66,
    "name": "Betty-1",
    "sex": "f",
    "skins": 1,
  },
  Object {
    "age": 8.13,
    "ageLastShaved": 8,
    "milk": 335.66,
    "name": "Betty-2",
    "sex": "f",
    "skins": 1,
  },
  Object {
    "age": 9.63,
    "ageLastShaved": 9.5,
    "milk": 277.16,
    "name": "Betty-3",
    "sex": "f",
    "skins": 1,
  },
]
`;

exports[`HerdService 1 - calculates expected values for 14 days 1`] = `
Array [
  Object {
    "age": 4.14,
    "ageLastShaved": 4.13,
    "milk": 529.27,
    "name": "Betty-1",
    "sex": "f",
    "skins": 2,
  },
  Object {
    "age": 8.14,
    "ageLastShaved": 8,
    "milk": 361.27,
    "name": "Betty-2",
    "sex": "f",
    "skins": 1,
  },
  Object {
    "age": 9.64,
    "ageLastShaved": 9.5,
    "milk": 298.27,
    "name": "Betty-3",
    "sex": "f",
    "skins": 1,
  },
]
`;

exports[`HerdService 2 - calculates expected values for 200 days 1`] = `
Array [
  Object {
    "age": 6,
    "ageLastShaved": 5.89,
    "milk": 7003,
    "name": "Betty-1",
    "sex": "f",
    "skins": 15,
  },
  Object {
    "age": 10,
    "ageLastShaved": 9.93,
    "milk": 4603,
    "name": "Betty-2",
    "sex": "f",
    "skins": 12,
  },
  Object {
    "age": 10,
    "ageLastShaved": 9.86,
    "milk": 1038.25,
    "name": "Betty-3",
    "sex": "f",
    "skins": 3,
  },
]
`;
